compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3135,36 +3122,23 @@ generateS390CompareAndBranchOpsHelper(TR::Node * node, TR::CodeGenerator * cg, T
    else
       {
       TR_S390BinaryCommutativeAnalyser temp(cg);
-
-      // FIXME: Why are we doing this business with the secondChild's type
-      // here? How can the second child's type ever be different from the
-      // first child's type?
-      // Answer:The only case where it can be different is in the compressed refs
-      // build. an aload of a class pointer is 32-bits only.
-      //
-      TR::DataType childType = secondChild->getDataType();
-      if (childType == TR::Address)
+      TR_ASSERT_FATAL(firstChild->getDataType() == secondChild->getDataType(), "Data types of both operands should match for generating compare and branch.");
+      // On 64-Bit platform with compressed references, it is possible that one of the child is loading a class from the object while other can be
+      // loading it from the temp stack slot in this case, VFT load should be evaluated into register where it loads 4 bytes with lowest byte cleared and
+      // zero extend it so we will have a VFT pointer in 64-Bit register so when it should generate instruction to compare 8 bytes.
+      TR::DataType childType = TR::Compiler->target.is32Bit() && firstChild->getDataType() == TR::Address ? TR::Int32 : firstChild->getDataType();
+      if (TR::Compiler->target.is64Bit())
          {
-         isUnsignedCmp = true;
-         if (cg->comp()->target().is64Bit() && !cg->isCompressedClassPointerOfObjectHeader(secondChild))
-            {
-            childType = TR::Int64;
-            }
-         else
-            {
-            childType = TR::Int32;
-            }
+         if (firstChild->getDataType() == TR::Address && cg->isCompressedClassPointerOfObjectHeader(firstChild))
+            cg->evaluate(firstChild);
+         if (secondChild->getDataType() == TR::Address && cg->isCompressedClassPointerOfObjectHeader(secondChild))
+            cg->evaluate(secondChild);
          }
 
-
-      if (childType == TR::Aggregate)
-         {
-         TR_ASSERT( 0, "unexpected data size for aggregate compare\n");
-         }
       switch (childType)
          {
          case TR::Int64:
-         case TR::Address: // FIXME: impossible to get here with TR::Address
+         case TR::Address:
```

Comment:
I'm not the biggest fan of this because it leans towards 64-bit as being the address type and does something special for 32-bit. We don't have to make this change here but I'll just note that a better way to do this `switch` could be to simply switch on the data type size rather than the type itself. This would map `TR::Address` to 8 on 64-bit and 4 on 32-bit respectively.

Reply:
Changed in https://github.com/eclipse/omr/pull/4763/commits/72d58c3cd5edcd89e97671d490393e1b8e01f9db